Linux: Fix: connectHandler after adv.Start

nckrss opened the pull request:
In Linux the Adapter.connectHandler is never called after Advertisement.Start.

This fix preserves the existing interfaces and adds D-Bus signals to handle connect/disconnect events once the Peripheral is advertising.

- Fix Linux Advertise.Start to call a connectHandler, if defined
- Update inline documentation for Adapter.SetConnectHandler
- Update README usage example

Potentially fixes #360 and #354.

nckrss committed:
Linux: Fix: connectHandler after adv.Start

In Linux the Adapter.connectHandler is never called after Advertisement.Start.

This fix preserves the existing interfaces and adds D-Bus signals to handle
connect/disconnect events once the Peripheral is advertising.

- Fix Linux Advertise.Start to call the connectHandler, if set
- Update inline documentation for Adapter.SetConnectHandler
- Update README usage example

Signed-off-by: Nick Ross <nick.ross@intel.com>

deadprogram commented on README.md:
```diff
@@ -22,6 +22,8 @@ This example shows a central that scans for peripheral devices and then displays
 package main
 
 import (
+	"fmt"
```
We try to avoid use of `fmt` unless really needed, in order to reduce size of the binary especially on embedded devices. That is the reason the examples are this way.

nckrss replied:
No worries, I will revert this change.

nckrss replied:
Reverted in b53217e

deadprogram commented on gap_linux.go:
```diff
@@ -496,3 +533,65 @@ func (a *Adapter) SetRandomAddress(mac MAC) error {
 }
+
+// SetAdapterAlias sets the alias of the adapter in BlueZ.
+func (a *Adapter) SetAdapterAlias(alias string) error {
```
Does this function really need to be exported? Since there is no equivalent on other platforms, we try to avoid it. Since it is not used in the PR itself, is this something you need for your specific application?

nckrss replied:
I'm fine with removing it if you prefer, the `LocalName` is available to Centrals that support and use extended advertisements.

For background context:

Given:

```go
// ...
if err := p.adv.Configure(bluetooth.AdvertisementOptions{
	LocalName: "TinyGo",
	ServiceUUIDs: []bluetooth.UUID{ServiceUUID},
}); err != nil {
	return fmt.Errorf("configure advertisement: %w", err)
}

```

in BlueZ 5.79 and 6.14 kernel the `LocalName` in `bluetooth.AdvertisementOptions` sets the name as part of the Extended Advertising Data, however it does not change the adapter alias which is currently being sent in the initial advertising packet, so the function `SetAdapterAlias` explicitly sets the adapter alias.

In `gap_linux.go` I could modify `Advertisement.Configure` to use the `LocalName` to also set the adapter `Alias`.  Since this is host stack specific, I don't think other changes (e.g.  adding  `AdvertisementOptions.AliasName`) makes sense, so if there are any changes I suggest keeping changes internal to the linux stack.

What do you prefer?

deadprogram replied:
> In gap_linux.go I could modify Advertisement.Configure to use the LocalName to also set the adapter Alias. Since this is host stack specific, I don't think other changes (e.g. adding AdvertisementOptions.AliasName) makes sense, so if there are any changes I suggest keeping changes internal to the linux stack.

That seems like a reasonable approach.

nckrss committed:
Revert README examples to use println

Signed-off-by: Nick Ross <nick.ross@intel.com>

nckrss committed:
Add resource cleanup in example

deadprogram commented on README.md:
```diff
@@ -67,6 +68,14 @@ func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	adapter.SetConnectHandler(func(device bluetooth.Device, connected bool){
+		if connected {
+			println("device connected:", device.Address.String(), device.RSSI, device.LocalName())
```
```suggestion
			println("device connected:", device.Address.String())
```

Neither `device.RSSI` nor `device.LocalName()` exist, which causes an error on my machine:

```
examples/advertisement/main.go:18:65: device.RSSI undefined (type bluetooth.Device has no field or method RSSI)
examples/advertisement/main.go:18:78: device.LocalName undefined (type bluetooth.Device has no field or method LocalName)
```

deadprogram replied:
After I changed the line of code to the above, it did work for me.

```shell
$ go run ./examples/advertisement/
advertising...
device connected: 6A:20:3A:EC:0A:B6
```

nckrss replied:
I corrected the example, and I also added `getDeviceProperties` used in handling `org.freedesktop.DBus.Properties.PropertiesChanged` signals and parsing known [`org.bluez.Device1` properties](https://github.com/luetzel/bluez/blob/master/doc/device-api.txt).

In this PR `getDeviceProperties` is just used for the `Address` property, however the `select` can be extended in subsequent PRs.  For example, the `Device` struct could add:

```go
Name    string
UUIDs   []UUID  // 128-bit UUIDs of the available services on the device
```

In local testing, when a Peripheral is advertising and a Central connects, the services available from the Central are available from D-Bus and can be captured through this function.

The `select` could then add, for example:

```go
case "Name":
	if name, ok := v.Value().(string); ok {
		d.Name = name
	}
case "UUIDs":
	if uuids, ok := v.Value().([]string); ok {
		for _, uuidStr := range uuids {
			if uuid, err := ParseUUID(uuidStr); err == nil {
				d.UUIDs = append(d.UUIDs, uuid)
			}
		}
	}
}
```

I believe these additional fields are too far out of scope for this PR, so I omitted them for now.

deadprogram commented on README.md:
```diff
@@ -67,6 +68,14 @@ func main() {
 	must("enable BLE stack", adapter.Enable())
 
+	ctx, cancel := context.WithCancel(context.Background())
+	adapter.SetConnectHandler(func(device bluetooth.Device, connected bool){
```
Actually, perhaps you want to change the entire handler to this:

```go
	adapter.SetConnectHandler(func(device bluetooth.Device, connected bool) {
		if connected {
			println("device connected:", device.Address.String())
			return
		}

		println("device disconnected:", device.Address.String())
		cancel()
	})
```

I think that perhaps better captures your intent here? Please let me know what you think.

nckrss replied:
I believe your example is more complete because it exits after a full lifecycle of device connection and disconnection.  I will update the README.

nckrss committed:
Advertising, signal handling, and README corrections

- Add alias update on adveristing
- Extend property handling for signals
- Correct README example

Signed-off-by: Nick Ross <nick.ross@intel.com>

deadprogram commented on gap_linux.go:
```diff
@@ -496,3 +544,74 @@ func (a *Adapter) SetRandomAddress(mac MAC) error {
+			}
+
+			// Call the connect handler if the Connected property has changed.
+			if connected, ok := changes["Connected"].Value().(bool); ok {
```
I wonder if this is not completely correct? Seems like the first time that a device connects, since there is no value, there is no call to the connect handler. The following disconnect, and subsequent connect/disconnects, do work correctly.

See the following output:

```
$ go run ./examples/advertisement/
advertising...
device disconnected: 59:5E:7D:4D:4B:09

$ go run ./examples/advertisement/
advertising...
device connected: 59:5E:7D:4D:4B:09
device disconnected: 59:5E:7D:4D:4B:09
```

What do you think?

nckrss replied:
I pushed a correction that seems to have corrected the behavior, although your confirmation is appreciated! I also updated the example.

deadprogram replied:
Confirmed that it is now working as expected. Thanks!

nckrss committed:
Add connectHandler to advertisement example

Signed-off-by: Nick Ross <nick.ross@intel.com>

nckrss committed:
Invoke handler on device initial connection

Signed-off-by: Nick Ross <nick.ross@intel.com>

deadprogram commented:
Thanks for working on this @nckrss now squash/merging.
